crates/shielded_token/src/conversion.rs

```diff
@@ -85,6 +85,28 @@ where
     Ok((checked!(10u128 ^ precision_denom)?, denomination))
 }
 
+/// Get the balance of the given token at the MASP address that is eligble to
+/// receive rewards.
+fn get_masp_dated_balance<S, TransToken>(
+    storage: &mut S,
+    token: &Address,
+) -> Result<Amount>
+where
+    S: StorageWrite + StorageRead,
+    TransToken: trans_token::Keys + trans_token::Read<S>,
+{
+    use crate::read_undated_balance;
+    let masp_addr = MASP;
+
+    // total locked amount in the Shielded pool
+    let total_tokens_in_masp =
+        TransToken::read_balance(storage, token, &masp_addr)?;
+    // Since dated and undated tokens are stored together in the pool, subtract
+    // the latter to get the dated balance
+    let masp_dated_balance = read_undated_balance(storage, token)?;
+    Ok(checked!(total_tokens_in_masp - masp_dated_balance)?)
```

Comment:
It should always hold that `masp_undated_balance <= total_tokens_in_masp`. If this checked subtraction fails, we are in trouble. I'd prefer unwrapping `checked!()`.
```suggestion
    Ok(checked!(total_tokens_in_masp - masp_dated_balance).unwrap())
```

Reply:
True, an underflow here would most likely mean that `masp_dated_balance` is in a very bad state. That being said, the policy followed for protocol code in this codebase seems to be to use checked arithmetic and propagate/handle arithmetic errors even if we believe or can prove that there will never be underflow. In the case that I have made a logical error in this PR, it would probably be better to just have the reward distribution mechanism fail and not distribute anything than to crash the whole node.